### PR TITLE
Replace flake8 with ruff in pre-commit and ci linting

### PR DESCRIPTION
**Pull request by mraspaud**
This PR uses ruff instead of flake8 and mccabe.

It also add news rules for linting, so I'm expecting this to fail and start a discussion about what rules to use.

When that is done, I will fix satpy accordingly.

As first draft, here what I propose:
- port of previously used flake8 and plugins used in satpy
	- [flake8-builtins](https://docs.astral.sh/ruff/rules/#flake8-builtins-a)
	- [flake8-bugbear](https://docs.astral.sh/ruff/rules/#flake8-bugbear-b)
	- [pydocstyle](https://docs.astral.sh/ruff/rules/#pydocstyle-d)
	- [mccabe](https://docs.astral.sh/ruff/rules/#mccabe-c90)
	- [flake8-debugger](https://docs.astral.sh/ruff/rules/#flake8-debugger-t10)

- new checks
	- [pycodestyle](https://docs.astral.sh/ruff/rules/#pycodestyle-e-w)
	- [pyflakes](https://docs.astral.sh/ruff/rules/#pycodestyle-e-w)
	- [isort](https://docs.astral.sh/ruff/rules/#isort-i)
	- [flake8-bandit](https://docs.astral.sh/ruff/rules/#flake8-bandit-s)
	- [flake8-pyteststyle](https://docs.astral.sh/ruff/rules/#flake8-pytest-style-pt)
	- [flake8-quotes](https://docs.astral.sh/ruff/rules/#flake8-quotes-q)
	- [flake8-tidyimports](https://docs.astral.sh/ruff/rules/#flake8-tidy-imports-tid)
	- [pep8-naming](https://docs.astral.sh/ruff/rules/#pep8-naming-n)

I'm also open to adding more checks, the full list is here: 
https://docs.astral.sh/ruff/rules/

### CODE REVIEW

**Comment by pnuu** [comment]
What triggers this change? Is it necessary to switch over for some reason, or is `ruff` simply better in some way?

**Comment by djhoese** [comment]
I am OK with adopting ruff, but I think we could drop the lint/style checks in the CI in favor of pre-commit.ci doing it all.

Also...adopt black?

**Comment by mraspaud** [comment]
> I am OK with adopting ruff, but I think we could drop the lint/style checks in the CI in favor of pre-commit.ci doing it all.

Ok, I can do that

> Also...adopt black?

Black is not a linter, this is about linting :grin:

**Comment by djhoese** [comment]
Sure, a separate PR. But it would be run in pre-commit so it isn't way out of the realm of conversation here and is something that would automatically fix many syntax/style errors complained about here.

**Comment by mraspaud** [comment]
Indeed, I was just trying to avoid the question as I don't like black and I would rather not have it in satpy, which may not be a popular opinion.

I would like this PR to be about ruff, but we can make an issue about black to discuss the pros and cons of it if you like.

**Review by djhoese — APPROVED** [review]
I vote for removing the linting job in CI, but otherwise I'm good with whatever rules you want to apply. Can we tell ruff what version(s) of python to target so you don't have to have the B905 rule ignored?

**Comment by gerritholl** [comment]
Looks good, but let's merge this after and not before the new release.

File: pyproject.toml [hunk 1]
```diff
@@ -12,3 +12,18 @@ skip_gitignore = true
 default_section = "THIRDPARTY"
 known_first_party = "satpy"
 line_length = 120
+
+[tool.ruff]
+select = ["E", "W", "F", "I", "D", "S", "B", "A", "PT", "Q", "TID", "C90", "T10", "N"]
```
Comment (gerritholl) [inline pyproject.toml:17]: Looks like magic ☺

```suggestion
[tool.ruff]
# See https://docs.astral.sh/ruff/rules/
select = ["E", "W", "F", "I", "D", "S", "B", "A", "PT", "Q", "TID", "C90", "T10", "N"]
```

**Review by gerritholl — APPROVED** [review]

**Comment by mraspaud** [comment]
> Looks good, but let's merge this after and not before the new release.

Will do!

**Comment by mraspaud** [comment]
@djhoese @gerritholl ok, I think have to stop working on this, because it's way too time consuming for my health atm. I did not add all the rules I planned to, but the ones we have current in main are implemented, and even a couple new ones.

If you want to review this, I highly recommend you check the changes by commit, since they should be consistent with one new rule per commit. Especially the one for pytest-style, which is a mouthful.

**Comment by djhoese** [comment]
There are a lot of test failures.

**Pull request merged**